### PR TITLE
Scaling of a Discrete Log Language

### PR DESCRIPTION
**Pull request by ycscaly**
This PR implements the Scaling of Discrete Log Language

### CODE REVIEW

File: src/scaling_of_discrete_log.rs [hunk 1]
```diff
@@ -0,0 +1,856 @@
...
+///
+/// However knowledge-soundness proofs are group and encryption scheme dependent, and thus we can
+/// only assure security for groups and encryption schemes for which we know how to prove it.
+///
+/// In the paper, we have proved it for any prime known-order group; so it is safe to use with a
+/// `PrimeOrderGroupElement`.
+///
+/// In regards to additively homomorphic encryption schemes, we proved it for `paillier`.
+#[derive(Clone, Serialize, Deserialize, PartialEq, Debug, Eq)]
+pub struct Language<
```
Comment (dolevmu) [inline src/scaling_of_discrete_log.rs:36]: Let's add like we did with after the audit, in the comments a mapping between the variable names in the code and in the paper. Also let's describe the language definition here mathematically as well.

File: src/scaling_of_discrete_log.rs [hunk 2]
```diff
@@ -0,0 +1,856 @@
...
+            PLAINTEXT_SPACE_SCALAR_LIMBS,
+            EncryptionKey,
+        >::new(
+            language_public_parameters.ciphertext,
+            language_public_parameters
+                .encryption_scheme_public_parameters
+                .ciphertext_space_public_parameters(),
+        )?;
+
+        // no mask needed, as we don't need circuit privacy.
```
Comment (dolevmu) [inline src/scaling_of_discrete_log.rs:160]: Now that I'm reading it, I think we should be a bit more careful/clear:
```suggestion
        // No masking of the plaintext is needed, as we don't need secure function evaluation.
        // However, we do want to re-randomize the ciphertext when doing the scalar multiplication, to ensure circuit privacy against an adversary that does not hold the private key, that is, the centralised party A.
```
I think the same comment should also be changed in EncDH if I remember correctly.

**Review by dolevmu — APPROVED** [review]
Looks good to me, but it would be easier for people to verify your code if you add the annotations.

**Review by ofirfriedman42 — APPROVED** [review]

File: src/scaling_of_discrete_log.rs [hunk 3]
```diff
@@ -0,0 +1,856 @@
...
+use serde::{Deserialize, Serialize};
+
+use crate::{language::DecomposableWitness, EnhanceableLanguage};
+
+/// Scaling of a Discrete Log Maurer Language
+///
+/// SECURITY NOTICE:
+/// This language implicitly assumes that the plaintext space of the encryption scheme and the
+/// scalar group coincide (same exponent). Using generic encryption schemes is permitted if and only
+/// if we use this language in its enhanced form, i.e. `EnhancedLanguage`.
```
Comment (zeevmoney) [inline src/scaling_of_discrete_log.rs:18]:
```suggestion
/// if we use this language in its enhanced form, i.e., `EnhancedLanguage`.
```

File: src/scaling_of_discrete_log.rs [hunk 4]
```diff
@@ -0,0 +1,856 @@
...
+///
+/// SECURITY NOTICE (2):
+/// Furthermore, even when using `EnhancedLanguage`, note that ENC_DH proves a correct computation
+/// that is not a secure function evaluation. That is, the result is not safe to decrypt, as it does
```
Comment (zeevmoney) [inline src/scaling_of_discrete_log.rs:22]:
```suggestion
/// Furthermore, even when using `EnhancedLanguage`, note that `ENC_DH` proves a correct computation
/// that is not a secure function evaluation. That is, the result is unsafe to decrypt, as it does
```

File: src/scaling_of_discrete_log.rs [hunk 5]
```diff
@@ -0,0 +1,856 @@
...
+/// not hide the number of arithmetic reductions mod q. For secure function evaluation, use
+/// `DComEval` (enhanced) language. Because correctness and zero-knowledge is guaranteed for any
+/// group and additively homomorphic encryption scheme in this language, we choose to provide a
+/// fully generic implementation.
+///
+/// However knowledge-soundness proofs are group and encryption scheme dependent, and thus we can
+/// only assure security for groups and encryption schemes for which we know how to prove it.
```
Comment (zeevmoney) [inline src/scaling_of_discrete_log.rs:29]:
```suggestion
/// However, knowledge-soundness proofs are group and encryption scheme-dependent, and thus we can
/// only ensure security for groups and encryption schemes for which we know how to prove it.
```

File: src/scaling_of_discrete_log.rs [hunk 6]
```diff
@@ -0,0 +1,856 @@
...
+///
+/// In the paper, we have proved it for any prime known-order group; so it is safe to use with a
+/// `PrimeOrderGroupElement`.
+///
+/// In regards to additively homomorphic encryption schemes, we proved it for `paillier`.
```
Comment (zeevmoney) [inline src/scaling_of_discrete_log.rs:34]:
```suggestion
/// Regarding additively homomorphic encryption schemes, we proved it for `Paillier`.
```

File: src/scaling_of_discrete_log.rs [hunk 7]
```diff
@@ -0,0 +1,856 @@
...
+    homomorphic_encryption::CiphertextSpaceGroupElement<
+        PLAINTEXT_SPACE_SCALAR_LIMBS,
+        EncryptionKey,
+    >,
+    GroupElement,
+>;
+
+/// The Public Parameters of the Scaling of a Discrete Log Maurer Language.
+/// The `lower_bound` of `ciphertext` should be verified independently,
+/// e.g. by verifying (and following) a sequence of enhanced proofs over the homomorphic
```
Comment (zeevmoney) [inline src/scaling_of_discrete_log.rs:75]:
```suggestion
/// e.g., by verifying (and following) a sequence of enhanced proofs over the homomorphic
```

File: src/scaling_of_discrete_log.rs [hunk 8]
```diff
@@ -0,0 +1,856 @@
...
+            .statement_space_public_parameters)
+            .into();
+
+        group_public_parameters
+    }
+}
+
+pub trait WitnessAccessors<
+    PlaintextSpaceGroupElement: group::GroupElement,
+    RandomnessSpaceGroupElement: group::GroupElement,
```
Comment (zeevmoney) [inline src/scaling_of_discrete_log.rs:429]:
```suggestion
    PlaintextSpaceGroupElement: GroupElement,
    RandomnessSpaceGroupElement: GroupElement,
```
no need to fix prefix everywhere.

**Review by zeevmoney — CHANGES_REQUESTED** [review]
see comments, also, there is a lot of duplicate code taken from encryption of discrete log, can you move it to functions so we won't have to maintain the same code in two different places?

**Comment by ycscaly** [comment]
Annotations should be added across the board, let's do that in a separate PR

**Pull request merged**